Fix lookup for current Thread in async signal handler

janvorli opened the pull request:
The current `CheckActivationSafePoint` uses thread local storage to
get the current Thread instance. But this function is called from
async signal handler (the activation signal handler) and it is not
allowed to access TLS variables there because the access can allocate
and if the interrupted code was running in an allocation code, it
could crash.
There was no problem with this since .NET 1.0, but a change in the
recent glibc version has broken this. We've got reports of crashes
in this code due to the reason mentioned above.

This change introduces an async safe mechanism for accessing the
current Thread instance from async signal handlers. It uses a
segmented array that can grow, but never shrink. Entries for
threads are added when runtime creates a thread / attaches to an
external thread and removed when the thread dies.

The check for safety of the activation injection was further enhanced
to make sure that the ScanReaderLock is not taken. In cases it would 
need to be taken, we just reject the location.

Since NativeAOT is subject to the same issue, the code to maintain the
thread id to thread instance map is placed to the minipal and shared
between coreclr and NativeAOT.

Closes #121581

janvorli committed:
Fix lookup for current Thread in async signal handler

The current CheckActivationSafePoint uses thread local storage to
get the current Thread instance. But this function is called from
async signal handler (the activation signal handler) and it is not
allowed to access TLS variables there because the access can allocate
and if the interrupted code was running in an allocation code, it
could crash.
There was no problem with this since .NET 1.0, but a change in the
recent glibc version has broken this. We've got reports of crashes
in this code due to the reason mentioned above.

This change introduces an async safe mechanism for accessing the
current Thread instance from async signal handlers. It uses a
segmented array that can grow, but never shrink. Entries for
threads are added when runtime creates a thread / attaches to an
external thread and removed when the thread dies.

Closes #121581

janvorli committed:
Add support for NativeAOT

jkotas commented:
> There was no problem with this since .NET 1.0, but a change in the recent glibc version has broken this

Have you been able to trace down the change? I think we just got lucky that it has not showed up on the radar earlier.

jkotas commented on src/native/minipal/thread.c:
```diff
@@ -0,0 +1,130 @@
+    struct ThreadSegment* pNext;
+};
+
+static struct ThreadSegment *s_pAsyncSafeThreadMapHead = NULL;
```
I think about src\minipal as general-purpose stuff that should be abstracting the platform. I know we have a few exceptions that are not really a platform abstraction - but they are still very general-purpose utilities.

This data structure looks very specific to how the thread suspension is implemented in JIT (regular CoreCLR) and NAOT runtimes. I think about it as an implementation detail of the thread suspension. Ideally, the whole implementation of thread suspension would be shared between JIT and NAOT - we are not there yet.

> cpuid/cpufeatures

BTW: cpuid/cpufeatures is used in crossgen2/ilc that is distinct from the runtimes.

> src\coreclr\minipal is the place for that facility

coreclr\minipal has the messier and less general-purpose PAL abstractions that are specific to the JIT runtime. It is not used by anything else. We tend to be moving bits of src\coreclr\pal that are too difficult to cleanup completely, https://github.com/dotnet/runtime/blob/main/src/coreclr/minipal/dn-stdio.h#L15 is a great example.

> coreclr\runtime is to unify asm code between them.

It is meant to be used for more than just the .asm files. It has CachedInterfaceDispatch.cpp/CachedInterfaceDispatch.h today. It would be nice to move the other .h/cpp files that are shared between JIT and NAOT runtime too. These files are cherry-picked from VM directory in a hacky way currently that does not look great, e.g.: https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/Runtime/gcinfodecoder.cpp.

jkotas commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -5731,16 +5731,18 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
-    Thread *pThread = GetThreadNULLOk();
+    Thread *pThread = GetThreadAsyncSafe();
+    _ASSERTE(pThread != NULL);
 
```
This looks like valid feedback. We can be seeing a signal sent by some other component (e.g. second NAOT runtime). We should forward the signal to the next component in that case.

Could you please do some ad-hoc stress testing to make sure that it works well? We should not be crashing, and we should not be eating signals that other components may expect.

janvorli committed:
Move the implementation to src/coreclr/runtime

AaronRobinsonMSFT commented on src/native/minipal/thread.c:
```diff
@@ -0,0 +1,130 @@
+            {
+                // Another thread added the segment first
+                free(pNewSegment);
+                pNewSegment = *ppSegment;
```
Instead of reaching back up and accessing a potentially changing value, this API writes the non-equal failure value back into `pExpected`. I think the more correct operation would be `pNewSegment = *pExpected;`

janvorli commented on src/coreclr/nativeaot/Runtime/threadstore.cpp:
```diff
@@ -352,6 +365,13 @@ EXTERN_C RuntimeThreadLocals* RhpGetThread()
+#ifdef TARGET_UNIX
+Thread * ThreadStore::GetCurrentThreadIfAvailableAsyncSafe()
+{
+    return (Thread*)FindThreadInAsyncSafeMap(minipal_get_current_thread_id_no_cache());
```
The cached variant uses TLS. We cannot use TLS, as preventing access to TLS is what's this PR is all about.

janvorli committed:
Fix build break

janvorli committed:
Reflect PR feedback

janvorli commented on src/native/minipal/thread.h:
done

jkotas commented on src/native/minipal/thread.c:
```diff
@@ -0,0 +1,130 @@
+#include "utils.h"
+#include "thread.h"
+
+#ifdef TARGET_UNIX
```
I agree. I do not think we want to pay this overhead on Windows.

janvorli committed:
More PR feedback

janvorli committed:
Fix another build break

jkotas commented on src/native/minipal/thread.h:
```diff
@@ -4,6 +4,8 @@
 #ifndef HAVE_MINIPAL_THREAD_H
 #define HAVE_MINIPAL_THREAD_H
 
+#ifndef TARGET_WINDOWS
```
```suggestion
#ifndef HOST_WINDOWS
```

janvorli replied:
Done

janvorli committed:
Reflect PR feedback and fix windows build break

janvorli commented on src/native/minipal/thread.h:
```diff
@@ -118,8 +138,11 @@ static inline int minipal_set_thread_name(pthread_t thread, const char* name)
 }
 #endif // extern "C"
 
+#endif // !TARGET_WINDOWS
```
Done

janvorli commented on src/coreclr/runtime/asyncsafethreadmap.cpp:
```diff
@@ -0,0 +1,130 @@
+        for (size_t i = 0; i < MAX_THREADS_IN_SEGMENT; i++)
+        {
+            size_t index = (startIndex + i) % MAX_THREADS_IN_SEGMENT;
+            // Use acquire to synchronize with release in insert_thread_to_async_safe_map
```
Fixed

janvorli committed:
One more build break

janvorli commented:
Looking at the NativeAOT handler, there is actually also a call to the GetCurrentThreadIfAvailable in the Thread::HijackCallback that , so the NativeAOT fix is not complete yet.

janvorli committed:
Forward activation signal to previous handler

On coreclr, forward the activation signal to a previously registered
handler (if any). On NativeAOT, that was already happening.

Also on NativeAOT, pass the Thread we get using the async safe mechanism
to the Thread::HijackCallback so that it doesn't try to get it itself.

jkotas commented on src/coreclr/nativeaot/Runtime/threadstore.cpp:
```diff
@@ -143,6 +145,14 @@ void ThreadStore::AttachCurrentThread(bool fAcquireThreadStoreLock)
+#if defined(TARGET_UNIX) && !defined(TARGET_WASM)
+    if (!InsertThreadIntoAsyncSafeMap(pAttachingThread->m_threadId, pAttachingThread))
+    {
+        ASSERT_UNCONDITIONALLY("Failed to insert thread into async-safe map due to OOM.");
```
```suggestion
        PalPrintFatalError("\nFailed to insert thread into async-safe map due to out of memory.\n");
```

jkotas commented on src/coreclr/vm/threads.cpp:
```diff
@@ -370,7 +383,27 @@ void SetThread(Thread* t)
+#if defined(TARGET_UNIX) && !defined(TARGET_WASM)
+        if (!InsertThreadIntoAsyncSafeMap(t->GetOSThreadId64(), t))
+        {
+            // TODO: can we handle this OOM more gracefully?
```
We can - if it is one of our threads. However, I do not think it is worth it. The allocation of the thread statics above can crash due to OOM too and I do not think there is a good way to recover from it either. Recovering from corner-case OOMs is just not a thing on Unix.

jkotas commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -5731,16 +5731,17 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
+        ExecutionManager::IsManagedCodeNoLock(ip);
 
-    if (!isActivationSafePoint)
+    if (!isActivationSafePoint && pThread != NULL)
```
Ah, it will be cleared later in HandleSuspensionForInterruptedThread - that's better.

jkotas reviewed (approved):
Thank you!

janvorli committed:
Apply suggestions from code review

Update comment and error messages

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

gbalykov commented:
Thanks! Do you plan to backport this to .net 8?
